docs/zq/language.md

```diff
@@ -1,341 +1,2091 @@
+then the abbreviation is treated as having an implied `yield` operator, e.g.,
+```
+{s:lower(s)}
+```
+is shorthand for
+```
+yield {s:lower(s)}
+```
+When operator names are omitted, `search` has precedence over `yield`, so
+```
+foo
+```
+is interpreted as a search for the string "foo" rather than a yield of
+the implied record field named `foo`.
 
-Generally speaking, a [flow graph](https://en.wikipedia.org/wiki/Directed_acyclic_graph)
-defines a directed acyclic graph (DAG) composed
-of data sources and operator nodes.  The Zed syntax leverages "fat arrows",
-i.e., `=>`, to indicate the start of a parallel path and terminates each
-parallel path with a semicolon.
+Another common query pattern involves adding or mutating fields of records
+where the input is presumed to be a sequence of records.
+The [put operator](operators/put.md#operator) provides this mechanism and the `put`
+keyword is implied by the mutator syntax `:=`, which is used in Zed when an
+input record field is modified, as compared to `=` which is used in constant
+and variable assignments.  For example, the operation
+```
+put y:=2*x+1
+```
+can be expressed simply as
+```
+y:=2*x+1
+```
+When composing long-form queries that are shared via Brim or managed in Github,
+it is best practice to include all operator names in the Zed source text.
 
-A data path can be split with the `split` operator as in
+In summary, if no operator name is given, the implied operator is determined
+from the operator-less source text, in the order given, as follows:
+* If the text can be interpreted as a search expression, then the operator is `search`.
+* If the text can be interpreted as a boolean expression, then the operator is `where`.
+* If the text can be interpreted as one or more field assignments, then the operator is `put`.
+* If the text can be interpreted as an aggregation, then the operator is `summarize`.
+* If the text can be interpreted as an expression, then the operator is `yield`.
+* Otherwise, the text causes a compile-time error.
+
+When in doubt, you can always check what the compiler is doing under the hood
+by running `zq` with the `-C` flag to print the parsed query in "canonical form", e.g.,
+```mdtest-command
+zq -C foo
+zq -C 'is(<foo>)'
+zq -C 'count()'
+zq -C '{a:x+1,b:y-1}'
+zq -C 'a:=x+1,b:=y-1'
 ```
-from PoolOne | split (
-  => op1 | op2 | ... ;
-  => op1 | op2 | ... ;
-) | merge ts | ...
+produces
+```mdtest-output
+search foo
+where is(<foo>)
+summarize
+    count()
+yield {a:x+1,b:y-1}
+put a:=x+1,b:=y-1
 ```
 
-> **Note:** Adding `merge` to the Zed language is still a work in progress
-> ([zed/2906](https://github.com/brimdata/zed/issues/2906)).
+## 3. Const Statements
 
-Or multiple pools can be accessed and, for example, joined:
+Constants may be defined and assigned to a symbolic name with the syntax
 ```
-from (
-  pool PoolOne => op1 | op2 | ...
-  pool PoolTwo => op1 | op2 | ...
-) | join on key=key | ...
+const <id> = <expr>
```

Comment:
To avoid confusion with Zed type values, I think we should adopt a convention other than angle brackets for syntax examples like this. Caps or italics, maybe?

Reply:
I agree but this is used everywhere.  I created #3270